File: skills/academic-figure/templates/rf-0018-survival-risk-table/r/plot.R
#!/usr/bin/env Rscript

# Plot supplied survival step coordinates and supplied risk counts using base R.

all_args <- commandArgs(trailingOnly = FALSE)
file_arg <- sub("^--file=", "", all_args[grep("^--file=", all_args)])
script_dir <- if (length(file_arg)) dirname(normalizePath(file_arg[[1]], mustWork = TRUE)) else getwd()
template_dir <- dirname(script_dir)
demo_steps <- file.path(template_dir, "demo", "demo_survival_steps_seed18.csv")
demo_risk <- file.path(template_dir, "demo", "demo_risk_table_seed18.csv")
demo_annotations <- file.path(template_dir, "demo", "demo_annotations_seed18.csv")

defaults <- list(
  steps = "", risk = "", annotations = "",
  output_prefix = file.path(getwd(), "survival_supplied_r"),
  title = "Supplied survival step coordinates", x_label = "Time", dpi = 320L
)

parse_args <- function(args) {
  result <- defaults
  aliases <- c(steps = "steps", risk = "risk", annotations = "annotations", `output-prefix` = "output_prefix", title = "title", `x-label` = "x_label", dpi = "dpi")
  index <- 1L
  while (index <= length(args)) {
    token <- args[[index]]
    if (!startsWith(token, "--")) stop(sprintf("Unknown argument: %s", token), call. = FALSE)
    stripped <- substring(token, 3L)
    if (grepl("=", stripped, fixed = TRUE)) {
      pieces <- strsplit(stripped, "=", fixed = TRUE)[[1]]
      key <- pieces[[1]]; value <- paste(pieces[-1], collapse = "=")
    } else {
      key <- stripped; index <- index + 1L
      if (index > length(args)) stop(sprintf("Missing value for --%s", key), call. = FALSE)
      value <- args[[index]]
    }
    if (!key %in% names(aliases)) stop(sprintf("Unknown option: --%s", key), call. = FALSE)
    result[[aliases[[key]]]] <- value
    index <- index + 1L
  }
  if (xor(nzchar(result$steps), nzchar(result$risk))) stop("Provide --steps and --risk together.", call. = FALSE)
  if (!nzchar(result$steps)) {
    result$steps <- demo_steps; result$risk <- demo_risk
    if (!nzchar(result$annotations)) result$annotations <- demo_annotations
  }
  result$dpi <- suppressWarnings(as.integer(result$dpi))
  if (is.na(result$dpi) || result$dpi < 150L) stop("--dpi must be at least 150.", call. = FALSE)
  result
}

stop_contract <- function(message) stop(sprintf("Input validation failed: %s", message), call. = FALSE)
ordered_unique <- function(values) values[!duplicated(values)]

read_character_csv <- function(path, required, label) {
  if (!file.exists(path) || dir.exists(path)) stop_contract(sprintf("%s does not exist: %s", label, path))
  data <- tryCatch(
    utils::read.csv(path, stringsAsFactors = FALSE, check.names = FALSE, colClasses = "character", na.strings = character()),
    error = function(error) stop_contract(conditionMessage(error))
  )
  names(data) <- trimws(names(data))
  if (any(!nzchar(names(data))) || anyDuplicated(names(data))) stop_contract(sprintf("%s headers must be nonblank and unique.", label))
  missing <- setdiff(required, names(data))
  if (length(missing)) stop_contract(sprintf("%s is missing columns: %s", label, paste(missing, collapse = ", ")))
  if (nrow(data)) {
    keep <- apply(data, 1L, function(row) any(nzchar(trimws(row))))
    data <- data[keep, , drop = FALSE]
  }
  if (!nrow(data)) stop_contract(sprintf("%s contains no data rows.", label))
  data
}

finite_number <- function(text, field, row_number) {
  value <- trimws(text)
  if (!nzchar(value)) stop_contract(sprintf("Row %d: '%s' must not be blank.", row_number, field))
  number <- suppressWarnings(as.numeric(value))
  if (is.na(number) || !is.finite(number)) stop_contract(sprintf("Row %d: '%s' must be finite numeric.", row_number, field))
  number
}

metadata_columns <- function(data) {
  data$status <- if ("data_status" %in% names(data)) toupper(trimws(data$data_status)) else rep("", nrow(data))
  data$seed <- if ("simulation_seed" %in% names(data)) trimws(data$simulation_seed) else rep("", nrow(data))
  data
}

read_steps <- function(path) {
  data <- metadata_columns(read_character_csv(path, c("time", "estimate", "curve_id"), "Step CSV"))
  data$curve_id <- trimws(data$curve_id)
  data$time_num <- NA_real_; data$estimate_num <- NA_real_
  for (index in seq_len(nrow(data))) {
    row_number <- index + 1L
    if (!nzchar(data$curve_id[[index]])) stop_contract(sprintf("Row %d: curve_id must not be blank.", row_number))
    data$time_num[[index]] <- finite_number(data$time[[index]], "time", row_number)
    data$estimate_num[[index]] <- finite_number(data$estimate[[index]], "estimate", row_number)
    if (data$time_num[[index]] < 0) stop_contract(sprintf("Row %d: time must be >= 0.", row_number))
    if (data$estimate_num[[index]] < 0 || data$estimate_num[[index]] > 1) stop_contract(sprintf("Row %d: estimate must lie in [0, 1].", row_number))
  }
  curves <- ordered_unique(data$curve_id)
  for (curve in curves) {
    subset <- data[data$curve_id == curve, , drop = FALSE]
    if (nrow(subset) < 2L) stop_contract(sprintf("Curve '%s' needs at least two coordinates.", curve))
    if (any(diff(subset$time_num) <= 0)) stop_contract(sprintf("Curve '%s' must appear in strictly increasing time order.", curve))
    if (any(diff(subset$estimate_num) > 1e-12)) stop_contract(sprintf("Curve '%s' increases; supplied survival steps must be non-increasing.", curve))
  }
  list(data = data, curves = curves)
}

read_risk <- function(path, steps) {
  data <- metadata_columns(read_character_csv(path, c("time", "curve_id", "n_at_risk"), "Risk CSV"))
  data$curve_id <- trimws(data$curve_id)
  data$time_num <- NA_real_; data$n_num <- NA_integer_
  known <- steps$curves
  for (index in seq_len(nrow(data))) {
    row_number <- index + 1L
    curve <- data$curve_id[[index]]
    if (!curve %in% known) stop_contract(sprintf("Risk row %d: unknown curve_id '%s'.", row_number, curve))
    data$time_num[[index]] <- finite_number(data$time[[index]], "time", row_number)
    curve_times <- steps$data$time_num[steps$data$curve_id == curve]
    if (data$time_num[[index]] < min(curve_times) || data$time_num[[index]] > max(curve_times)) stop_contract(sprintf("Risk row %d: time is outside the supplied curve range.", row_number))
    count <- suppressWarnings(as.integer(trimws(data$n_at_risk[[index]])))
    if (is.na(count) || count < 0L || as.character(count) != trimws(data$n_at_risk[[index]])) stop_contract(sprintf("Risk row %d: n_at_risk must be a nonnegative integer.", row_number))
    data$n_num[[index]] <- count
  }
  if (anyDuplicated(paste(data$curve_id, data$time_num, sep = "\r"))) stop_contract("Risk curve_id/time pairs must be unique.")
  grids <- lapply(known, function(curve) {
    subset <- data[data$curve_id == curve, , drop = FALSE]
    if (!nrow(subset)) stop_contract(sprintf("Risk table is missing curve '%s'.", curve))
    if (any(diff(subset$time_num) <= 0)) stop_contract(sprintf("Risk rows for '%s' must appear in increasing time order.", curve))
    subset$time_num
  })
  if (any(vapply(grids[-1], function(grid) !identical(grid, grids[[1]]), logical(1)))) stop_contract("All curves must use the same ordered risk-table time grid.")
  list(data = data, times = grids[[1]])
}

read_annotations <- function(path, steps) {
  if (!nzchar(path)) return(data.frame())
  data <- metadata_columns(read_character_csv(path, c("time", "curve_id", "label"), "Annotation CSV"))
  data$curve_id <- trimws(data$curve_id); data$label <- trimws(data$label); data$time_num <- NA_real_
  for (index in seq_len(nrow(data))) {
    row_number <- index + 1L; curve <- data$curve_id[[index]]
    if (!curve %in% steps$curves) stop_contract(sprintf("Annotation row %d: unknown curve_id.", row_number))
    if (!nzchar(data$label[[index]])) stop_contract(sprintf("Annotation row %d: label must not be blank.", row_number))
    data$time_num[[index]] <- finite_number(data$time[[index]], "time", row_number)
    limits <- range(steps$data$time_num[steps$data$curve_id == curve])
    if (data$time_num[[index]] < limits[[1]] || data$time_num[[index]] > limits[[2]]) stop_contract(sprintf("Annotation row %d: time is outside the curve range.", row_number))
  }
  if (anyDuplicated(paste(data$curve_id, data$time_num, sep = "\r"))) stop_contract("Annotation curve_id/time pairs must be unique.")
  data
}

metadata_note <- function(data_frames) {
  statuses <- unlist(lapply(data_frames, function(data) if (nrow(data)) data$status else character()))
  seeds <- unlist(lapply(data_frames, function(data) if (nrow(data)) data$seed else character()))
  unique_statuses <- ordered_unique(statuses[nzchar(statuses)])
  unique_seeds <- ordered_unique(seeds[nzchar(seeds)])
  if (length(unique_statuses) > 1L || length(unique_seeds) > 1L) stop_contract("data_status and simulation_seed must be constant across all files.")
  if (identical(unique_statuses, "SIMULATED")) {
    if (any(statuses != "SIMULATED") || any(!nzchar(seeds))) stop_contract("Every simulated row must declare the same fixed seed.")
    seed <- suppressWarnings(as.integer(unique_seeds[[1]]))
    if (is.na(seed) || seed <= 0L || as.character(seed) != unique_seeds[[1]]) stop_contract("simulation_seed must be a positive integer.")
    return(sprintf("SIMULATED DEMONSTRATION DATA · fixed seed %d", seed))
  }
  if (length(unique_seeds)) stop_contract("simulation_seed is only valid with SIMULATED data.")
  "SOURCE-SUPPLIED / PRECOMPUTED DATA"
}

step_value <- function(steps, curve, time) {
  subset <- steps[steps$curve_id == curve, , drop = FALSE]
  index <- findInterval(time, subset$time_num)
  subset$estimate_num[[max(1L, index)]]
}

draw_figure <- function(steps, risk, annotations, curves, risk_times, note, options) {
  background <- "#FBFAF7"; ink <- "#20282C"; muted <- "#647078"; grid_color <- "#E5E2DC"
  palette <- c("#287D9B", "#D96B35", "#4F8A5B", "#8D65A8", "#B68A1F", "#5875B5")
  colors <- stats::setNames(rep(palette, length.out = length(curves)), curves)
  graphics::layout(matrix(c(1, 2), nrow = 2), heights = c(3.4, max(1.65, 0.55 * length(curves))))
  graphics::par(oma = c(3.0, 0.5, 4.5, 0.5), bg = background, fg = ink, family = "sans")
  all_times <- steps$time_num; span <- diff(range(all_times)); padding <- max(0.02 * span, 0.1)

  graphics::par(mar = c(0.7, 4.7, 0.7, 1.0))
  graphics::plot.new(); graphics::plot.window(xlim = range(all_times) + c(-padding, padding), ylim = c(-0.015, 1.025), xaxs = "i", yaxs = "i")
  x_ticks <- graphics::axTicks(1); y_ticks <- graphics::axTicks(2)
  graphics::abline(v = x_ticks, h = y_ticks, col = grid_color, lwd = 0.65)
  for (curve in curves) {
    subset <- steps[steps$curve_id == curve, , drop = FALSE]
    graphics::lines(subset$time_num, subset$estimate_num, type = "s", col = colors[[curve]], lwd = 1.8)
    graphics::points(subset$time_num, subset$estimate_num, pch = 21, bg = colors[[curve]], col = "white", cex = 0.78, lwd = 0.45)
  }
  if (nrow(annotations)) {
    for (index in seq_len(nrow(annotations))) {
      row <- annotations[index, ]; y <- step_value(steps, row$curve_id, row$time_num)
      graphics::segments(row$time_num, y, row$time_num + 0.18 * padding, y + 0.045, col = colors[[row$curve_id]], lwd = 0.7)
      graphics::text(row$time_num + 0.22 * padding, y + 0.055, row$label, adj = c(0, 0), cex = 0.68, col = colors[[row$curve_id]])
    }
  }
  graphics::axis(2, las = 1, cex.axis = 0.78, col = "#6D7A80")
  graphics::box(bty = "l", col = "#6D7A80")
  graphics::mtext("Supplied survival estimate", side = 2, line = 3.1, cex = 0.88)
  graphics::legend("bottomleft", legend = curves, col = unname(colors[curves]), lty = 1, lwd = 1.7, pch = 21, pt.bg = unname(colors[curves]), bty = "n", horiz = length(curves) <= 4L, cex = 0.76)

  graphics::par(mar = c(3.2, 4.7, 1.45, 1.0))
  graphics::plot.new(); graphics::plot.window(xlim = range(all_times) + c(-padding, padding), ylim = c(length(curves) + 0.5, 0.5), xaxs = "i", yaxs = "i")
  for (row_index in seq_along(curves)) if (row_index %% 2L == 1L) graphics::rect(min(all_times) - padding, row_index - 0.5, max(all_times) + padding, row_index + 0.5, col = "#F1EFE9", border = NA)
  graphics::abline(v = risk_times, col = grid_color, lwd = 0.55)
  for (row_index in seq_along(curves)) {
    curve <- curves[[row_index]]
    for (time in risk_times) {
      count <- risk$n_num[risk$curve_id == curve & risk$time_num == time]
      graphics::text(time, row_index, count, cex = 0.76)
    }
  }
  graphics::axis(1, at = risk_times, labels = format(risk_times, trim = TRUE), cex.axis = 0.78, col = "#6D7A80")
  graphics::axis(2, at = seq_along(curves), labels = curves, las = 1, tick = FALSE, cex.axis = 0.78)
  graphics::box(bty = "l", col = "#6D7A80")
  graphics::mtext(options$x_label, side = 1, line = 2.4, cex = 0.88)
  graphics::mtext("Supplied number at risk", side = 3, line = 0.45, adj = 0, cex = 0.88, font = 2)

  graphics::mtext(options$title, side = 3, outer = TRUE, line = 2.8, adj = 0.02, cex = 1.75, font = 2)
  graphics::mtext(note, side = 3, outer = TRUE, line = 1.35, adj = 0.02, cex = 0.86, col = muted)
  graphics::mtext("SUPPLIED/PRECOMPUTED coordinates, risk counts, and optional annotations; no Kaplan-Meier fit, confidence interval, or log-rank test is computed.", side = 3, outer = TRUE, line = 0.25, adj = 0.02, cex = 0.72, col = muted)
  graphics::mtext("Post-step convention: each estimate is held until the next supplied coordinate. Statistical interpretation remains upstream.", side = 1, outer = TRUE, line = 1.35, adj = 0.02, cex = 0.69, col = muted)
}

options <- parse_args(commandArgs(trailingOnly = TRUE))
steps <- read_steps(options$steps)
risk <- read_risk(options$risk, steps)
annotations <- read_annotations(options$annotations, steps)
note <- metadata_note(list(steps$data, risk$data, annotations))
width <- min(17, max(9, 7.2 + 0.62 * length(risk$times) + 0.03 * max(nchar(steps$curves))))
height <- max(7.0, 5.2 + 0.42 * length(steps$curves))
prefix <- file.path(dirname(options$output_prefix), tools::file_path_sans_ext(basename(options$output_prefix)))
dir.create(dirname(prefix), recursive = TRUE, showWarnings = FALSE)
png_path <- paste0(prefix, ".png"); pdf_path <- paste0(prefix, ".pdf")
grDevices::png(png_path, width = width, height = height, units = "in", res = options$dpi, bg = "#FBFAF7")
draw_figure(steps$data, risk$data, annotations, steps$curves, risk$times, note, options); grDevices::dev.off()
grDevices::pdf(pdf_path, width = width, height = height, bg = "#FBFAF7", onefile = TRUE)
draw_figure(steps$data, risk$data, annotations, steps$curves, risk$times, note, options); grDevices::dev.off()
message(sprintf("Validated %d supplied step coordinates, %d risk counts, %d supplied annotations, %d curves", nrow(steps$data), nrow(risk$data), nrow(annotations), length(steps$curves)))
message(sprintf("Data status: %s", note))
message(sprintf("PNG: %s", normalizePath(png_path, mustWork = TRUE)))
message(sprintf("PDF: %s", normalizePath(pdf_path, mustWork = TRUE)))

File: skills/academic-figure/templates/rf-0018-survival-risk-table/python/plot.py
#!/usr/bin/env python3
"""Plot supplied survival step coordinates with a supplied risk table."""

from __future__ import annotations

import argparse
import bisect
import csv
import math
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, Iterable, List, Optional, Sequence, Tuple

import matplotlib.pyplot as plt


SCRIPT_DIR = Path(__file__).resolve().parent
TEMPLATE_DIR = SCRIPT_DIR.parent
DEMO_STEPS = TEMPLATE_DIR / "demo" / "demo_survival_steps_seed18.csv"
DEMO_RISK = TEMPLATE_DIR / "demo" / "demo_risk_table_seed18.csv"
DEMO_ANNOTATIONS = TEMPLATE_DIR / "demo" / "demo_annotations_seed18.csv"
DEFAULT_OUTPUT_PREFIX = Path.cwd() / "survival_supplied_python"
COLORS = ("#287D9B", "#D96B35", "#4F8A5B", "#8D65A8", "#B68A1F", "#5875B5")
BACKGROUND = "#FBFAF7"
INK = "#20282C"
MUTED = "#647078"
GRID = "#E5E2DC"


class ContractError(ValueError):
    pass


@dataclass(frozen=True)
class StepRow:
    line: int
    time: float
    estimate: float
    curve_id: str
    status: str
    seed: str


@dataclass(frozen=True)
class RiskRow:
    line: int
    time: float
    curve_id: str
    n_at_risk: int
    status: str
    seed: str


@dataclass(frozen=True)
class AnnotationRow:
    line: int
    time: float
    curve_id: str
    label: str
    status: str
    seed: str


def ordered_unique(values: Iterable[str]) -> List[str]:
    seen = set()
    output: List[str] = []
    for value in values:
        if value not in seen:
            seen.add(value)
            output.append(value)
    return output


def finite_number(text: str, field: str, line: int) -> float:
    value = text.strip()
    if not value:
        raise ContractError(f"Line {line}: '{field}' must not be blank.")
    try:
        number = float(value)
    except ValueError as exc:
        raise ContractError(f"Line {line}: '{field}' must be numeric.") from exc
    if not math.isfinite(number):
        raise ContractError(f"Line {line}: '{field}' must be finite.")
    return number


def csv_rows(path: Path, required: Sequence[str], label: str) -> List[Tuple[int, Dict[str, str]]]:
    if not path.is_file():
        raise ContractError(f"{label} does not exist: {path}")
    output: List[Tuple[int, Dict[str, str]]] = []
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle)
        if reader.fieldnames is None:
            raise ContractError(f"{label} has no header row.")
        headers = [header.strip() for header in reader.fieldnames]
        if len(headers) != len(set(headers)):
            raise ContractError(f"{label} headers must be unique.")
        missing = set(required) - set(headers)
        if missing:
            raise ContractError(f"{label} is missing columns: {', '.join(sorted(missing))}")
        for line, raw in enumerate(reader, start=2):
            row = {key: str(value or "").strip() for key, value in raw.items()}
            if all(not value for value in row.values()):
                continue
            output.append((line, row))
    if not output:
        raise ContractError(f"{label} contains no data rows.")
    return output


def read_steps(path: Path) -> Tuple[List[StepRow], List[str]]:
    rows: List[StepRow] = []
    for line, row in csv_rows(path, ("time", "estimate", "curve_id"), "Step CSV"):
        curve_id = row["curve_id"]
        if not curve_id:
            raise ContractError(f"Line {line}: curve_id must not be blank.")
        time = finite_number(row["time"], "time", line)
        estimate = finite_number(row["estimate"], "estimate", line)
        if time < 0:
            raise ContractError(f"Line {line}: time must be >= 0.")
        if not 0 <= estimate <= 1:
            raise ContractError(f"Line {line}: estimate must lie in [0, 1].")
        rows.append(StepRow(line, time, estimate, curve_id, row.get("data_status", "").upper(), row.get("simulation_seed", "")))
    curves = ordered_unique(row.curve_id for row in rows)
    for curve in curves:
        subset = [row for row in rows if row.curve_id == curve]
        if len(subset) < 2:
            raise ContractError(f"Curve {curve!r} needs at least two supplied step coordinates.")
        for previous, current in zip(subset, subset[1:]):
            if current.time <= previous.time:
                raise ContractError(
                    f"Curve {curve!r} must appear in strictly increasing time order; "
                    f"line {current.line} follows time {previous.time:g}."
                )
            if current.estimate > previous.estimate + 1e-12:
                raise ContractError(
                    f"Curve {curve!r} increases from {previous.estimate:g} to "
                    f"{current.estimate:g}; supplied survival steps must be non-increasing."
                )
    return rows, curves


def read_risk(path: Path, steps: Sequence[StepRow], curves: Sequence[str]) -> Tuple[List[RiskRow], List[float]]:
    rows: List[RiskRow] = []
    seen = set()
    known = set(curves)
    limits = {
        curve: (
            min(row.time for row in steps if row.curve_id == curve),
            max(row.time for row in steps if row.curve_id == curve),
        )
        for curve in curves
    }
    for line, row in csv_rows(path, ("time", "curve_id", "n_at_risk"), "Risk CSV"):
        curve = row["curve_id"]
        if curve not in known:
            raise ContractError(f"Risk line {line}: unknown curve_id {curve!r}.")
        time = finite_number(row["time"], "time", line)
        if time < 0 or not limits[curve][0] <= time <= limits[curve][1]:
            raise ContractError(
                f"Risk line {line}: time must lie inside the supplied range for {curve!r}."
            )
        text = row["n_at_risk"]
        try:
            count = int(text)
        except ValueError as exc:
            raise ContractError(f"Risk line {line}: n_at_risk must be a nonnegative integer.") from exc
        if count < 0 or str(count) != text:
            raise ContractError(f"Risk line {line}: n_at_risk must be a nonnegative integer.")
        key = (curve, time)
        if key in seen:
            raise ContractError(f"Risk line {line}: duplicate curve_id/time pair.")
        seen.add(key)
        rows.append(RiskRow(line, time, curve, count, row.get("data_status", "").upper(), row.get("simulation_seed", "")))

    grids: List[List[float]] = []
    for curve in curves:
        subset = [row for row in rows if row.curve_id == curve]
        if not subset:
            raise ContractError(f"Risk table is missing curve {curve!r}.")
        times = [row.time for row in subset]
        if any(current <= previous for previous, current in zip(times, times[1:])):
            raise ContractError(f"Risk rows for {curve!r} must appear in strictly increasing time order.")
        grids.append(times)
    if any(grid != grids[0] for grid in grids[1:]):
        raise ContractError("All curves must use the same ordered risk-table time grid.")
    return rows, grids[0]


def read_annotations(
    path: Optional[Path], steps: Sequence[StepRow], curves: Sequence[str]
) -> List[AnnotationRow]:
    if path is None:
        return []
    output: List[AnnotationRow] = []
    seen = set()
    known = set(curves)
    limits = {
        curve: (
            min(row.time for row in steps if row.curve_id == curve),
            max(row.time for row in steps if row.curve_id == curve),
        )
        for curve in curves
    }
    for line, row in csv_rows(path, ("time", "curve_id", "label"), "Annotation CSV"):
        curve = row["curve_id"]
        label = row["label"]
        if curve not in known:
            raise ContractError(f"Annotation line {line}: unknown curve_id {curve!r}.")
        if not label:
            raise ContractError(f"Annotation line {line}: label must not be blank.")
        time = finite_number(row["time"], "time", line)
        if not limits[curve][0] <= time <= limits[curve][1]:
            raise ContractError(f"Annotation line {line}: time is outside the curve range.")
        key = (curve, time)
        if key in seen:
            raise ContractError(f"Annotation line {line}: duplicate curve_id/time pair.")
        seen.add(key)
        output.append(AnnotationRow(line, time, curve, label, row.get("data_status", "").upper(), row.get("simulation_seed", "")))
    return output


def metadata_note(groups: Sequence[Sequence[object]]) -> str:
    rows = [row for group in groups for row in group]
    statuses = ordered_unique(getattr(row, "status") for row in rows if getattr(row, "status"))
    seeds = ordered_unique(getattr(row, "seed") for row in rows if getattr(row, "seed"))
    if len(statuses) > 1 or len(seeds) > 1:
        raise ContractError("data_status and simulation_seed must be constant across all supplied files.")
    if statuses == ["SIMULATED"]:
        if any(getattr(row, "status") != "SIMULATED" or not getattr(row, "seed") for row in rows):
            raise ContractError("Every simulated row in every supplied file must declare the same fixed seed.")
        try:
            seed = int(seeds[0])
        except (IndexError, ValueError) as exc:
            raise ContractError("simulation_seed must be a positive integer.") from exc
        if seed <= 0 or str(seed) != seeds[0]:
            raise ContractError("simulation_seed must be a positive integer.")
        return f"SIMULATED DEMONSTRATION DATA · fixed seed {seed}"
    if seeds:
        raise ContractError("simulation_seed is only valid with SIMULATED data.")
    return "SOURCE-SUPPLIED / PRECOMPUTED DATA"


def step_value(steps: Sequence[StepRow], curve: str, time: float) -> float:
    subset = [row for row in steps if row.curve_id == curve]
    times = [row.time for row in subset]
    index = bisect.bisect_right(times, time) - 1
    return subset[max(index, 0)].estimate


def render(
    steps: Sequence[StepRow],
    risk: Sequence[RiskRow],
    annotations: Sequence[AnnotationRow],
    curves: Sequence[str],
    risk_times: Sequence[float],
    note: str,
    output_prefix: Path,
    title: str,
    x_label: str,
    dpi: int,
) -> Tuple[Path, Path]:
    longest_curve = max(len(curve) for curve in curves)
    width = min(17.0, max(9.0, 7.2 + 0.62 * len(risk_times) + 0.03 * longest_curve))
    height = max(6.4, 4.8 + 0.34 * len(curves))
    figure = plt.figure(figsize=(width, height), facecolor=BACKGROUND)
    grid = figure.add_gridspec(
        2, 1, height_ratios=(3.4, max(1.15, 0.38 * len(curves))),
        left=0.15, right=0.97, top=0.82, bottom=0.14, hspace=0.18
    )
    curve_axis = figure.add_subplot(grid[0, 0])
    risk_axis = figure.add_subplot(grid[1, 0], sharex=curve_axis)
    colors = {curve: COLORS[index % len(COLORS)] for index, curve in enumerate(curves)}

    for curve in curves:
        subset = [row for row in steps if row.curve_id == curve]
        curve_axis.step(
            [row.time for row in subset],
            [row.estimate for row in subset],
            where="post",
            color=colors[curve],
            linewidth=1.8,
            label=curve,
        )
        curve_axis.scatter(
            [row.time for row in subset],
            [row.estimate for row in subset],
            s=18, color=colors[curve], edgecolor="white", linewidth=0.45, zorder=3
        )
    for annotation in annotations:
        y = step_value(steps, annotation.curve_id, annotation.time)
        curve_axis.annotate(
            annotation.label,
            xy=(annotation.time, y),
            xytext=(5, 9),
            textcoords="offset points",
            fontsize=7.4,
            color=colors[annotation.curve_id],
            arrowprops={"arrowstyle": "-", "color": colors[annotation.curve_id], "linewidth": 0.7},
        )
    all_times = [row.time for row in steps]
    span = max(all_times) - min(all_times)
    padding = max(0.02 * span, 0.1)
    curve_axis.set_xlim(min(all_times) - padding, max(all_times) + padding)
    curve_axis.set_ylim(-0.015, 1.025)
    curve_axis.set_ylabel("Supplied survival estimate")
    curve_axis.grid(color=GRID, linewidth=0.65)
    curve_axis.set_axisbelow(True)
    curve_axis.tick_params(axis="x", labelbottom=False)
    curve_axis.legend(frameon=False, ncol=min(4, len(curves)), loc="upper right", fontsize=8.0)
    for spine in ("top", "right"):
        curve_axis.spines[spine].set_visible(False)

    risk_axis.set_ylim(len(curves) - 0.5, -0.5)
    risk_axis.set_yticks(range(len(curves)))
    risk_axis.set_yticklabels(curves, fontsize=8.2)
    risk_axis.set_xticks(risk_times)
    risk_axis.set_xlabel(x_label)
    risk_axis.set_title("Supplied number at risk", loc="left", fontsize=9.2, fontweight="bold", pad=8)
    for row_index in range(len(curves)):
        if row_index % 2 == 0:
            risk_axis.axhspan(row_index - 0.5, row_index + 0.5, color="#F1EFE9", zorder=0)
    risk_lookup = {(row.curve_id, row.time): row.n_at_risk for row in risk}
    for row_index, curve in enumerate(curves):
        for time in risk_times:
            risk_axis.text(time, row_index, str(risk_lookup[(curve, time)]), ha="center", va="center", fontsize=8.0, color=INK)
    risk_axis.grid(axis="x", color=GRID, linewidth=0.55)
    risk_axis.tick_params(axis="y", length=0, pad=8)
    for spine in ("top", "right", "left"):
        risk_axis.spines[spine].set_visible(False)

    figure.text(0.04, 0.975, title, ha="left", va="top", fontsize=18, fontweight="bold", color=INK)
    figure.text(0.04, 0.928, note, ha="left", va="top", fontsize=9.0, color=MUTED)
    figure.text(
        0.04, 0.890,
        "SUPPLIED/PRECOMPUTED coordinates, risk counts, and optional annotations; no Kaplan–Meier fit, confidence interval, or log-rank test is computed.",
        ha="left", va="top", fontsize=8.2, color=MUTED,
    )
    figure.text(
        0.04, 0.025,
        "Step convention: each estimate is held from its supplied time until the next supplied coordinate (post step). Statistical interpretation remains upstream.",
        ha="left", va="bottom", fontsize=7.6, color=MUTED,
    )
    prefix = output_prefix.expanduser().with_suffix("")
    prefix.parent.mkdir(parents=True, exist_ok=True)
    png_path = prefix.with_suffix(".png")
    svg_path = prefix.with_suffix(".svg")
    figure.savefig(png_path, dpi=dpi, bbox_inches="tight", facecolor=BACKGROUND)
    figure.savefig(svg_path, bbox_inches="tight", facecolor=BACKGROUND)
    plt.close(figure)
    return png_path, svg_path


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Plot supplied survival steps and risk counts without fitting a model.")
    parser.add_argument("--steps", type=Path)
    parser.add_argument("--risk", type=Path)
    parser.add_argument("--annotations", type=Path)
    parser.add_argument("--output-prefix", type=Path, default=DEFAULT_OUTPUT_PREFIX)
    parser.add_argument("--title", default="Supplied survival step coordinates")
    parser.add_argument("--x-label", default="Time")
    parser.add_argument("--dpi", type=int, default=320)
    args = parser.parse_args()
    if (args.steps is None) != (args.risk is None):
        parser.error("Provide --steps and --risk together.")
    if args.steps is None:
        args.steps = DEMO_STEPS
        args.risk = DEMO_RISK
        if args.annotations is None:
            args.annotations = DEMO_ANNOTATIONS
    if args.dpi < 150:
        parser.error("--dpi must be at least 150.")
    return args


def main() -> int:
    args = parse_args()
    try:
        steps, curves = read_steps(args.steps.expanduser())
        risk, risk_times = read_risk(args.risk.expanduser(), steps, curves)
        annotations = read_annotations(
            args.annotations.expanduser() if args.annotations else None, steps, curves
        )
        note = metadata_note((steps, risk, annotations))
        png_path, svg_path = render(
            steps, risk, annotations, curves, risk_times, note,
            args.output_prefix, args.title, args.x_label, args.dpi
        )
    except ContractError as exc:
        raise SystemExit(f"Input validation failed: {exc}") from exc
    print(
        f"Validated {len(steps)} supplied step coordinates, {len(risk)} risk counts, "
        f"{len(annotations)} supplied annotations, {len(curves)} curves"
    )
    print(f"Data status: {note}")
    print(f"PNG: {png_path.resolve()}")
    print(f"SVG: {svg_path.resolve()}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
